### PR TITLE
[iOS] Add `UITextInteraction` only when `inputDelegate` is nil

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
- Fixes https://github.com/flutter/flutter/issues/99652 
- Fixes https://github.com/flutter/flutter/issues/98590

This pull request removes the `UITextInteraction` initialization in `[FlutterTextInputView initWithOwner:]` previously added to `FlutterTextInputView` in https://github.com/flutter/engine/pull/26486. 

When a certain `UITextInteraction` is removed from the text input view, it (incorrectly?) sets `inputDelegate` to `nil` instead of `[UIKeyboardImpl activeInstance]`. This workaround adds a `UITextInteraction` to the view when we need `inputDelegate` but it's currently `nil` (see `setTextInputState`), so the `UITextInteraction`'s setup process sets `inputDelegate` to a `UITextInteractionInputDelegate` which eventually notifies the current active `UIKeyboardImpl` in its `UITextInputDelegate` implementation. And we remove the `UITextInteraction` when we finish using `inputDelegate`.

A few observations: 

- Floating cursor seems to be working correctly even without the `UITextInteraction` now. 
- It seems `inputDelegate` is set to nil when the last/root text interaction is removed from the view. So another (easier) workaround for the issue is to add our own `UITextInteraction` to the input view to make sure there's always one `UITextInteraction` installed, but that adds [unwanted visual artifacts](https://github.com/flutter/engine/pull/32881#issuecomment-1111052552) during scribble.
## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by jmagman — COMMENTED** [review]
Was it an accident? https://github.com/flutter/engine/pull/24224#issuecomment-897138832

That comment says it was "delayed" not removed so maybe it was an accident?

Can you confirm this change doesn't regress
> for some reason, the OS seems to be adding a grey shaded rectangle to the screen when using scribble to delete. I suspect it is related to autocorrect highlighting, but I'm not sure. So far I haven't been able to get rid of it. It does go away when tapping the screen again.

**Comment by LongCatIsLooong** [comment]
@fbcouch it looks like the Japanese romaji keyboard needs the `UITextInteraction` to work (my guess is it has to do with the swipe gestures used on those keyboards). Is there an alternative workaround for the scribble delete issue?

**Comment by fbcouch** [comment]
Hm. I didn't have another workaround, but – I notice you're just changing where the interaction gets set up. Is it possible to either detect the language the user is using and conditionally set up the UITextInteraction there, or add something to the framework that tells the engine that it needs that to be already available?

**Comment by LongCatIsLooong** [comment]
The romaji keyboard doesn't seem to call any of `UITextInteraction`'s public properties/methods when I use swipe gesture to enter. 
But `UITextInteraction` does set `UITextInput.inputDelegate` to a `UITextInteractionInputDelegate` I guess that's why it is needed, whenever the text is changed the delegate needs to be notified.

**Comment by fbcouch** [comment]
I wonder if we might be able to make use of https://developer.apple.com/documentation/uikit/uiscribbleinteraction/3566750-pencilinputexpected?language=objc somehow to conditionally set up the UITextInteraction...or, from reading your comment on the linked issue, is there a way to just not clear the `TextInput.inputDelegate`?

**Comment by LongCatIsLooong** [comment]
Did a bit testing and `UITextInteraction` seems to be longer needed for floating cursor to work with marked text. We should be able to remove it if it wasn't for https://github.com/flutter/flutter/issues/99652.

https://github.com/flutter/flutter/issues/99652 seems to be a UIKit bug. @fbcouch I have a (not so pleasant) workaround to reset the `inputDelegate` to the current active`UIKeyboardImpl` instance. I'll update the pull request in a bit.

**Comment by LongCatIsLooong** [comment]
- the [workaround](https://github.com/flutter/engine/pull/32881/commits/517743a781357f9a1ea318ce39bb3972944b752c) unfortunately doesn't work well when there's composing text, as it requires the keyboard to restart, and as a result the currently marked (composing) text will be lost. 
- assigning `inputDelegate` to `[UIKeyboardImpl activeInstance]` should work but that's a private class
- trying to retain the `UITextInteractionInputDelegate` and remove the `UITextInteration` from the view will result in `EXC_BAD_ACCESS`.

The easiest way out is to keep the `UITextInteraction` in `init` until the bug is fixed from UIKit. @fbcouch how badly would scribble be affected if `UITextInteraction` is set up in `init`?

**Comment by LongCatIsLooong** [comment]
I guess I could add a dummy text interaction when the input delegate is being set to 0. I'll try that tomorrow

**Comment by fbcouch** [comment]
> The easiest way out is to keep the `UITextInteraction` in `init` until the bug is fixed in UIKit. @fbcouch how badly would scribble be affected if `UITextInteraction` is set up in `init`?

I think the only issue is that there's a grey rectangle left over in some cases (maybe when deleting with scribble?) that goes away on the next interaction. I wonder if an alternative workaround would be to remove the interaction when a scribble interaction starts, then add it back when it ends?

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm [hunk 1]
```diff
@@ -905,7 +905,27 @@ - (void)setTextInputClient:(int)client {
   _hasPlaceholder = NO;
 }
 
+- (UITextInteraction*)textInteraction API_AVAILABLE(ios(13.0)) {
+  if (!_textInteraction) {
+    _textInteraction =
+        [[UITextInteraction textInteractionForMode:UITextInteractionModeEditable] retain];
+    _textInteraction.textInput = self;
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:912]: Does `_textInteraction.textInput` need to be nil'd out somewhere to break the retain cycle?

**Review by jmagman — APPROVED** [review]

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm (hunk 1 above)
Comment (LongCatIsLooong) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:912]: https://developer.apple.com/documentation/uikit/uitextinteraction/3255084-textinput?language=objc this is a weak property and the framework is compiled with ARC so I assume it doesn't increase the retain count?

Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:912]: 👍  Thanks for checking.

**Review by jmagman — APPROVED** [review]
LGTM

**Pull request merged**